Attach error to state.

egidijusz opened the pull request:
When error happens on transit - it's logged on Finitomata level and "gone" afterwards (pls, correct me if I'm wrong).
I think it would be useful to "attach" transit error to FSM state. So it can be accessed and acted upon at `on_failure` callback.

Don't hesitate to decline the pull request if you don't agree with this change :)

Commit:
Attach error to state.

When error happens on transit - it's logged on Finitomata level and "gone" afterwards.
This commit "attaches" transit error to FSM state. So it can be accessed and acted upon at `on_failure` callback.

am-kantox commented on lib/finitomata.ex:
```diff
@@ -29,9 +29,10 @@ defmodule Finitomata do
             timer: non_neg_integer(),
-            history: [Transition.state()]
+            history: [Transition.state()],
+            error: any()
```
We produce this error, so we are free to tighten the type as strictly as we want.

egidijusz replied:
I'm not sure we will always know the error type. 
It comes from `else` in `with` statement. So, to my understanding, it can be unhandled result from `with`, also it can be `RuntimeError`, `ArgumentError` and etc.

am-kantox replied:
Ah, indeed, sorry for the noise, disregard it, please.

am-kantox commented on lib/finitomata.ex:
```diff
@@ -442,20 +443,21 @@ defmodule Finitomata do
           end
         else
           err ->
+            state_with_error = struct(state, error: err)
```
`state` is already a struct, I’d prefer a more explicit
```suggestion
            state = %State{state | error: err}
```

Also, it’s still our old good state, no need to have a new name, _Elixir_ is not 100% _Erlang_.

am-kantox commented on lib/finitomata.ex:
```diff
@@ -442,20 +443,21 @@ defmodule Finitomata do
               event in @__config_soft_events__ ->
                 Logger.debug("[⚐⥯] transition softly failed " <> inspect(err))
-                {:noreply, state}
+                {:noreply, state_with_error}
```
If you accept the rebinding proposed above, this and the two changes below should go away.

am-kantox commented on test/support/modules.ex:
```diff
@@ -144,3 +144,24 @@ defmodule Finitomata.Test.Soft do
+  @impl Finitomata
+  def on_failure(_event, _payload, state) do
+    Logger.debug("[failure] " <> inspect(state.error))
+    :ok
```
`Logger.debug/1` already returns `:ok`, and this is a test, no need to be _as_ explicit.
```suggestion
```

am-kantox reviewed (changes requested):
> Don’t hesitate to decline the pull request if you don't agree with this change :)

I maintain 10+ libs since 2010, I never hesitate to do anything :)

---

Jokes aside, I ❤ the idea.

What we need to do besides keeping the error, would be either clean it up _immediately after_ `on_failure/3` invocation or keep errors as a keyword list `[{transition(), reason}] when reason: any()`. Or, a keyword list of `[{state(), {transition(), reason}}]`, I am open to any suggestion.

The trick here would be to now overflood this list, so I’d go with a (maybe later configurable) number of last errors, let’s say, three.

WDYT?

egidijusz committed:
Update lib/finitomata.ex

Co-authored-by: Aleksei Matiushkin <aleksei.matiushkin@kantox.com>

Commit:
Fix state naming.

egidijusz committed:
Update test/support/modules.ex

Co-authored-by: Aleksei Matiushkin <aleksei.matiushkin@kantox.com>

am-kantox commented on lib/finitomata.ex:
```diff
@@ -442,6 +443,7 @@ defmodule Finitomata do
           end
         else
           err ->
+            state = %State{state | error: err}
```
After another fresh look, I hesitate to say what would be better, to keep a stale error forever (current,) to clean it up right before line 420 or to pass it to `safe_on_failure/3` _only_.

Thoughts?

This should be clarified before we merge the code.

egidijusz replied:
I've commented below on "cleaning up the error" part.
With regards to passing it only to `safe_on_failure/3` - maybe even better would be to pass it to `safe_on_failure/4`? As in - add new attribute to `safe_on_failure` which is the `err`? So we can keep state cleaner (and even have an easier way to pattern match the error).
What do you think?

egidijusz replied:
Also, if we pass only to `on_failure` (be it `/3` or `/4`) - the replies for soft events (line 448) and `:continue` for ensure entry (line 455) will lack the error.

am-kantox reviewed (commented):
1 question to clarify is left.

egidijusz commented:
> > Don’t hesitate to decline the pull request if you don't agree with this change :)
> 
> I maintain 10+ libs since 2010, I never hesitate to do anything :)
> 
> Jokes aside, I heart the idea.
> 
> What we need to do besides keeping the error, would be either clean it up _immediately after_ `on_failure/3` invocation or keep errors as a keyword list `[{transition(), reason}] when reason: any()`. Or, a keyword list of `[{state(), {transition(), reason}}]`, I am open to any suggestion.
> 
> The trick here would be to now overflood this list, so I’d go with a (maybe later configurable) number of last errors, let’s say, three.
> 
> WDYT?

Hmm. I was assuming that `on_failure` is kinda the final _step_ before the final state (like a cleanup before going to [*] and shutting down the FSM). Because all other business logic "failures" are covered by diverging transition paths (like "process transaction" --> "insufficient credit"). Leaving `on_failure` to handle runtime errors, missing keys in maps/structs, library raises and etc.
It's how I am using it 😅
So in this case keeping list of `{transition(), reason}` would not be needed. We have `Transition.event()` passed to `on_failure` anyway.

But from your comment I understand now that it might not be the final state. Am I correct here? If yes - I think the `[{transition(), reason}]` is good way to approach it. 
With regards to cleaning up the error - I would keep it as is and leave the cleaning up part for `on_failure` implementation. So the user of Finitomata gets all the info about error and are able to do with it as they wish. 
What do you think?

egidijusz commented:
P.s. sorry, my replies are all over the place 😅

am-kantox commented:
> But from your comment I understand now that it might not be the final state. Am I correct here? 

Yes. It would be called every time when the transition has not succeeded, not even only due to an exception raised, but also in cases when `on_transition/4` callback returned crap, or unreachable `to` state, or whatever.

> I would […] leave the cleaning up part for `on_failure/3` implementation.

No way, this callback is not intended to modify the state, it’s `on_failure(…) :: :ok`.

> I think the `[{transition(), reason}]` is good way to approach it. 

See below, TL;DR: I think we don’t need to keep a tail, the latest one would be enough, as `{from, event, reason}`.

> maybe even better would be to pass it to `safe_on_failure/4`?

Yes, but it’s a library and you know what people hate most in library adoption? When it changes the API out of the blue. We might go the long path, deprecating but still supporting `on_failure/3` till `v1.0.0`, but I don’t see it’s worth it.

> we can keep state cleaner (and even have an easier way to pattern match the error)

In the first place, one is still able to pattern match the error with `%State{error: %MyException}` directly in `on_failure/3`. Also, the error is surely part of the state. That said, I’d keep the only last error, and I’d rename the struct key to `last_error`. Analyzing several last errors nevertheless must be considered an antipattern, and the consumer of the library should workaround this in their code anyway.

> Also, if we pass only to on_failure (be it /3 or /4) - the replies for soft events (line 448) and `:continue` for ensure entry (line 455) will lack the error.

After the third rethought, I’d keep it everywhere (all three clauses of `cond/1`.) After all, the respective handler might analyze the error in the _previously failed_ attempts and act respectively.

egidijusz commented:
> Yes. It would be called every time when the transition has not succeeded, not even only due to an exception raised, but also in cases when `on_transition/4` callback returned crap, or unreachable `to` state, or whatever.

Got it. This makes more sense now.

> No way, this callback is not intended to modify the state, it’s `on_failure(…) :: :ok`.

Yes, but I was talking specifically about the error itself. For example cleaning it up before persisting it to logging provider.

> After the third rethought, I’d keep it everywhere (all three clauses of `cond/1`.) After all, the respective handler might analyze the error in the _previously failed_ attempts and act respectively.

So to summarize (I agree with your arguments about breaking API, pattern matching the state struct and etc) - we keep it as it is now, part of the state (as a `last_error`). Yes?
And what about cleaning up?

am-kantox commented:
> we keep it as it is now, part of the state (as a `last_error`). Yes?

Almost, as a `%State{state | last_error: [state: state.current, event: event, error: err]}` _or_ as `%State{state | last_error: {state.current, event, err}}` which one you like more :)

---

I think we don’t need to clean it up once it’s called `last_error` there should not be any ambiguity.

Commit:
Change state.error to state.last_error.

`state.last_error` is a map now.
Additionally to error itself it holds state and event at which the error happened.

egidijusz commented:
> Almost, as a `%State{state | last_error: [state: state.current, event: event, error: err]}` _or_ as `%State{state | last_error: {state.current, event, err}}` which one you like more :)

I would prefer `last_error` to be a map. Tuple is too non descriptive and pattern matching on keywords require to match all the elements and their order.

I've updated PR with a map implementation. But I can change it to keyword/tuple if required :)

am-kantox commented on lib/finitomata.ex:
```diff
@@ -23,15 +23,20 @@ defmodule Finitomata do
 
+    @typedoc "The map that holds last error which happened on transition (at given state and event)."
+    @type last_error ::
+            %{state: Transition.state(), event: Transition.event(), error: any()} | nil
```
❤

am-kantox reviewed (approved):
:zap: :droplet: :zap: :gem: :zap: :blue_heart: :zap:

am-kantox commented:
> Package published to https://hex.pm/packages/finitomata/0.8.2
> (bd934fab494ad3d0a966cf46c74d557ab122d193ff4fce69d180e03d963f5993)
